### PR TITLE
hir: Make sure all `HirId`s have corresponding HIR `Node`s

### PR DESCRIPTION
**Pull request by petrochenkov**
And then remove `tcx.opt_hir_node(hir_id)` in favor of `tcx.hir_node(hir_id)`.

### CODE REVIEW

**Comment by compiler-errors** [comment]
this is great!

perf run just to check 🤷, but i'll honestly probably approve this even if there is a minimal perf hit

@bors try @rust-timer queue

File: compiler/rustc_metadata/src/rmeta/encoder.rs
Comment (compiler-errors) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:1391]: 🎉

File: src/tools/clippy/tests/ui/manual_map_option_2.stderr
Comment (compiler-errors) [inline src/tools/clippy/tests/ui/manual_map_option_2.stderr:24]: What happened here?

Comment (petrochenkov) [reply src/tools/clippy/tests/ui/manual_map_option_2.stderr:24]: This is likely due to introduction of `Node::Let`, it can now be observed in a chain of `HirId` parents, and clippy could rely on it not being in the chain in some place.
I spent some limited time searching for that place, but couldn't find it.

I planned to make a clippy issue for this and for `src/tools/clippy/tests/ui/shadow.stderr` as well, because it also looks questionable (and related to `Node::Let`).

File: compiler/rustc_ast_lowering/src/index.rs
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/index.rs:66]: Should we pre-fill `nodes` with `(0, Node::Err)`, and have the indexer overwrite it, avoiding the re-allocation?

File: compiler/rustc_ast_lowering/src/index.rs
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/index.rs:59]: Should we give the missing `HirId`?

File: compiler/rustc_hir/src/hir.rs
Comment (cjgillot) [inline compiler/rustc_hir/src/hir.rs:3505]: Do we really need to store that span? Could we return `DUMMY_SP` when needed?

File: compiler/rustc_hir/src/hir.rs
Comment (cjgillot) [inline compiler/rustc_hir/src/hir.rs:3357]: Should we both take and return by value?

File: compiler/rustc_hir/src/hir.rs
Comment (cjgillot) [inline compiler/rustc_hir/src/hir.rs:3503]:
```suggestion
    ArrayLenInfer(&'hir ArrayLen),
```
?

**Comment by petrochenkov** [comment]
TODO: Also need to check whether these `HirId`s could be removed instead, maybe they are not really necessary and were added by mistake.

File: compiler/rustc_hir/src/hir.rs
Comment (petrochenkov) [reply compiler/rustc_hir/src/hir.rs:3503]: The second half of `ArrayLen` does not correspond to the id, it has its own it.

This could be `ArrayLenInfer(&'hir ArrayLenInfer)` if the fields of `ArrayLen::Infer` were factored into a separate structure.

File: compiler/rustc_hir/src/hir.rs
Comment (petrochenkov) [reply compiler/rustc_hir/src/hir.rs:3505]: We could return `DUMMY_SP`, yes.
`index_hir` just has a close enough real span that can be used, so I used it (that makes `Node::span` always return real spans).

File: compiler/rustc_hir/src/hir.rs
Comment (petrochenkov) [reply compiler/rustc_hir/src/hir.rs:3357]: It's returned by reference because it's used by reference in `Node::Err`.

Comment (Nilstrieb) [reply compiler/rustc_hir/src/hir.rs:3357]: It should still return `Span`, we should minimize the places where `&Span` is used.

Comment (petrochenkov) [reply compiler/rustc_hir/src/hir.rs:3357]: `Node::Err` needs `&'hir Span` with `'hir` lifetime specifically, and this function provides it.
It cannot return `Span` by value, unless we use something else (like manual reimplementation of this same function), for giving a span to `Node::Err`.

Comment (Nilstrieb) [reply compiler/rustc_hir/src/hir.rs:3357]: Ah, you then need to create a reference. Makes sense.

**Comment by petrochenkov** [comment]
> TODO: Also need to check whether these `HirId`s could be removed instead, maybe they are not really necessary and were added by mistake.

The `HirId` in `Let` turned out to be unnecessary, other `HirId`s from this PR are used in more substantial ways.
Blocked on https://github.com/rust-lang/rust/pull/120428.
@rustbot blocked

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/120206#discussion_r1461195337 helped to identify one more class of stray `HirId`s in import list stems.
They weren't detected previously because they were always at the end of `nodes` lists.
This case is also addressed in the first commit now.
@rustbot ready

**Review by compiler-errors — APPROVED** [review]

**Comment by Nilstrieb** [comment]
perf neutral
@bors rollup=maybe

**Pull request merged**
